rfcs/0133-git-hashing.md

```diff
@@ -0,0 +1,317 @@
+
+3. We also support plans to decouple the layers of Nix further, and update our educational and marketing material to reflect it.
+   With Flakes and other post-2.0 features, the upper layer of Nix have gained an enormous amount of flexibility and sophistication.
+   RFCs like this show that the so-far more sleepy lower layers also have plenty of potential to gain sophistication too.
+
+   Embracing layering on technical, educational, communications, and managerial levels can scale our capacity to manage complexity and sophistication without the project growing out of control.
+   It will "divide and conquer" the project so the interfaces between each layer are still rigorously enforced preventing a combinatorial explosion in complexity.
+   That frees up "complexity budget" for project like this.
+
+   We plan on more formally proposing this next.
+
+## Git and Nix's file system data models do not entirely coincide
+
+Nix puts the permission info of a file (executable bit for now) with that file, whereas Git puts it with the name and hash in the directory.
+The practical effect of this discrepancy is that a root file (as opposed to directory) in Nix has permission info, but does not in Git.
+
+If we are trying to convert existing Nix data into Git, this is a problem.
+Assuming we treat "no permission bits" as meaning "non-executable", we will have a partial conversion that will fail on executable bare files.
+Tricks like always wrapping everything in a directory get around this, but then we have to be careful the directory is exactly as expected when "unwrapping" in the other direction.
+
+For now, we only focus on ingesting data *from* Git *to* Nix, and this side-steps the issue.
+That conversation is total (though not surjective), and so there is no problem for now.
+
+# Alternatives
+[alternatives]: #alternatives
+
+The dependency graph of steps can be sliced to save some for future work.
+For now they are all written together, but during the RFC meetings we will decide which steps (if any) to ratify now, and which steps to save for later.
+
+# Unresolved questions
+[unresolved]: #unresolved-questions
+
+None at this time.
+
+# Future work --- IPFS
+[future]: #future-work
+
+## Motivation
+
+### Binary Distribution
+
+IPFS is a potential solution leveraging content addressing for a P2P CDN.
+It out-of-the-box match's Nix's data model extremely well.
+
+### Source distribution and archival
+
+IPFS also supports git hashing, and so we also provide a good way for people and institutions to "pin" the sources they need, especially if those sources include private ones SWH won't have.
+Finally, per [Obsidian's bridging work](https://github.com/obsidiansystems/go-ipfs-swh-plugin), we have a 3 way integration between IPFS, SWH, and Nix.
+Data can be directly downloaded from SWH via HTTPS, or indirectly via IPFS, which can act as a CDN to not put as much load on SWH's servers.
+
+### Not just IPFS
```

Comment:
This may be too far off topic but since IPFS was mentioned I recall that we discussed that IPFS has a max block size whereas Git doesn't. Can you elaborate how this will affect our Git hashing implementation and eventually our IPFS implementation? It seems that IPFS is one of the major motivators of this work so it would be nice to have a rough idea of how this will work in the end.

Reply:
It won't work for larger files until that restriction is lifted, but there is a few ideas on how to do that.

Reply:
Collapsing IPFS threads. I hope even if we just were to do git hashing and not IPFS the normative part would still stand on it's own so don't need to resolve this completely now?